`HttpFields$Mutable$Wrapper.computeField()` incorrectly calls `onAddField()` when null is returned by `computeFn`

**Jetty version(s)**
12.0.8

**Jetty Environment**
ee10

**Description**

The `HttpFields$Mutable.computeField()` contract explicitly specifies that if the `computeFn` returns null, this means the field should be removed. Unfortunately, the default implementation of `computeField()` calls `put()` which incorrectly calls `Wrapper.onAddField()` with a null value.

When the protocol is HTTP 1.0, `ResponseUtils.ensureNotPersistent()` uses `computeField()` to remove the connection header, so that triggers the following exception:

```
java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.http.HttpField.getHeader()" because "field" is null
	at org.eclipse.jetty.ee10.servlet.ServletChannelState.onError(ServletChannelState.java:874)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handleException(ServletChannel.java:667)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:584)
	at org.eclipse.jetty.ee10.servlet.AsyncContentProducer.lambda$isReady$0(AsyncContentProducer.java:260)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.run(ContextHandler.java:1298)
	at org.eclipse.jetty.server.handler.ContextRequest$OnContextDemand.run(ContextRequest.java:74)
	at org.eclipse.jetty.util.thread.SerializedInvoker$Link.run(SerializedInvoker.java:191)
	at org.eclipse.jetty.server.internal.HttpConnection$DemandContentCallback.succeeded(HttpConnection.java:679)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:311)
	at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.http.HttpField.getHeader()" because "field" is null
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse$HttpFieldsWrapper.onAddField(ServletContextResponse.java:567)
	at org.eclipse.jetty.http.HttpFields$Mutable$Wrapper.put(HttpFields.java:1678)
	at org.eclipse.jetty.http.HttpFields$Mutable.computeField(HttpFields.java:1372)
	at org.eclipse.jetty.server.ResponseUtils.ensureNotPersistent(ResponseUtils.java:40)
	at org.eclipse.jetty.server.ResponseUtils.ensureConsumeAvailableOrNotPersistent(ResponseUtils.java:31)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:554)
	... 17 more
```

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
@@ -1386,7 +1386,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
         default Mutable computeField(HttpHeader header, BiFunction<HttpHeader, List<HttpField>, HttpField> computeFn)
         {
             Objects.requireNonNull(header);
-            HttpField result = computeFn.apply(header, stream().filter(f -> f.getHeader() == header).collect(Collectors.toList()));
+            HttpField result = computeFn.apply(header, stream().filter(f -> f.getHeader() == header).toList());
             return result != null ? put(result) : remove(header);
         }
 
@@ -1401,7 +1401,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
         default Mutable computeField(String name, BiFunction<String, List<HttpField>, HttpField> computeFn)
         {
             Objects.requireNonNull(name);
-            HttpField result = computeFn.apply(name, stream().filter(f -> f.is(name)).collect(Collectors.toList()));
+            HttpField result = computeFn.apply(name, stream().filter(f -> f.is(name)).toList());
             return result != null ? put(result) : remove(name);
         }
 

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
@@ -898,8 +898,8 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(String name, String value)
         {
-            if (value == null)
-                throw new IllegalArgumentException("null value");
+            Objects.requireNonNull(name);
+            Objects.requireNonNull(value);
             return add(new HttpField(name, value));
         }
 
@@ -914,6 +914,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(String name, long value)
         {
+            Objects.requireNonNull(name);
             return add(new HttpField.LongValueHttpField(name, value));
         }
 
@@ -928,6 +929,8 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(HttpHeader header, HttpHeaderValue value)
         {
+            Objects.requireNonNull(header);
+            Objects.requireNonNull(value);
             return add(header, value.toString());
         }
 
@@ -942,8 +945,8 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(HttpHeader header, String value)
         {
-            if (value == null)
-                throw new IllegalArgumentException("null value");
+            Objects.requireNonNull(header);
+            Objects.requireNonNull(value);
             return add(new HttpField(header, value));
         }
 
@@ -958,6 +961,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(HttpHeader header, long value)
         {
+            Objects.requireNonNull(header);
             return add(new HttpField.LongValueHttpField(header, value));
         }
 
@@ -969,6 +973,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable add(HttpField field)
         {
+            Objects.requireNonNull(field);
             ListIterator<HttpField> i = listIterator(size());
             i.add(field);
             return this;
@@ -1024,6 +1029,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable addCSV(HttpHeader header, String... values)
         {
+            Objects.requireNonNull(header);
             QuotedCSV existing = null;
             for (HttpField f : this)
             {
@@ -1051,6 +1057,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable addCSV(String name, String... values)
         {
+            Objects.requireNonNull(name);
             QuotedCSV existing = null;
             for (HttpField f : this)
             {
@@ -1078,6 +1085,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable addDateField(String name, long date)
         {
+            Objects.requireNonNull(name);
             add(name, DateGenerator.formatDate(date));
             return this;
         }
@@ -1107,6 +1115,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default void ensureField(HttpField field)
         {
+            Objects.requireNonNull(field);
             HttpHeader header = field.getHeader();
             // Is the field value multi valued?
             if (field.getValue().indexOf(',') < 0)
@@ -1138,6 +1147,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(HttpField field)
         {
+            Objects.requireNonNull(field);
             boolean put = false;
             ListIterator<HttpField> i = listIterator();
             while (i.hasNext())
@@ -1172,6 +1182,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(String name, String value)
         {
+            Objects.requireNonNull(name);
             if (value == null)
                 return remove(name);
             return put(new HttpField(name, value));
@@ -1188,6 +1199,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(HttpHeader header, HttpHeaderValue value)
         {
+            Objects.requireNonNull(header);
             if (value == null)
                 return remove(header);
             return put(new HttpField(header, value.toString()));
@@ -1204,6 +1216,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(HttpHeader header, String value)
         {
+            Objects.requireNonNull(header);
             if (value == null)
                 return remove(header);
             return put(new HttpField(header, value));
@@ -1243,6 +1256,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable putDate(HttpHeader name, long date)
         {
+            Objects.requireNonNull(name);
             return put(name, DateGenerator.formatDate(date));
         }
 
@@ -1258,6 +1272,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable putDate(String name, long date)
         {
+            Objects.requireNonNull(name);
             return put(name, DateGenerator.formatDate(date));
         }
 
@@ -1271,6 +1286,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(HttpHeader header, long value)
         {
+            Objects.requireNonNull(header);
             if (value == 0 && header == HttpHeader.CONTENT_LENGTH)
                 return put(HttpFields.CONTENT_LENGTH_0);
             return put(new HttpField.LongValueHttpField(header, value));
@@ -1286,6 +1302,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable put(String name, long value)
         {
+            Objects.requireNonNull(name);
             if (value == 0 && HttpHeader.CONTENT_LENGTH.is(name))
                 return put(HttpFields.CONTENT_LENGTH_0);
             return put(new HttpField.LongValueHttpField(name, value));
@@ -1369,7 +1386,9 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable computeField(HttpHeader header, BiFunction<HttpHeader, List<HttpField>, HttpField> computeFn)
         {
-            return put(computeFn.apply(header, stream().filter(f -> f.getHeader() == header).collect(Collectors.toList())));
+            Objects.requireNonNull(header);
+            HttpField result = computeFn.apply(header, stream().filter(f -> f.getHeader() == header).collect(Collectors.toList()));
+            return result != null ? put(result) : remove(header);
         }
 
         /**
@@ -1382,7 +1401,9 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable computeField(String name, BiFunction<String, List<HttpField>, HttpField> computeFn)
         {
-            return put(computeFn.apply(name, stream().filter(f -> f.is(name)).collect(Collectors.toList())));
+            Objects.requireNonNull(name);
+            HttpField result = computeFn.apply(name, stream().filter(f -> f.is(name)).collect(Collectors.toList()));
+            return result != null ? put(result) : remove(name);
         }
 
         /**
@@ -1393,6 +1414,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable remove(HttpHeader header)
         {
+            Objects.requireNonNull(header);
             Iterator<HttpField> i = iterator();
             while (i.hasNext())
             {
@@ -1430,6 +1452,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
          */
         default Mutable remove(String name)
         {
+            Objects.requireNonNull(name);
             for (ListIterator<HttpField> i = listIterator(); i.hasNext(); )
             {
                 HttpField f = i.next();
@@ -1658,6 +1681,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
             @Override
             public Mutable put(HttpField field)
             {
+                Objects.requireNonNull(field);
                 // rewrite put to ensure that removes are called before replace
                 int put = -1;
                 ListIterator<HttpField> i = _fields.listIterator();
@@ -1677,7 +1701,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
                 {
                     field = onAddField(field);
                     if (field != null)
-                        add(field);
+                        _fields.add(field);
                 }
                 else
                 {

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
@@ -1005,8 +1005,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
         default Mutable add(String name, List<String> list)
         {
             Objects.requireNonNull(name);
-            if (list == null)
-                throw new IllegalArgumentException("null list");
+            Objects.requireNonNull(list);
             if (list.isEmpty())
                 return this;
             if (list.size() == 1)

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
@@ -1004,7 +1004,7 @@ public class HttpFieldsTest
     public void testAddNullValueList()
     {
         HttpFields.Mutable fields = HttpFields.build();
-        assertThrows(IllegalArgumentException.class, () -> fields.add("name", (List<String>)null));
+        assertThrows(NullPointerException.class, () -> fields.add("name", (List<String>)null));
         assertThat(fields.size(), is(0));
         List<String> list = new ArrayList<>();
         fields.add("name", list);

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
@@ -1374,4 +1374,56 @@ public class HttpFieldsTest
         fields.ensureField(new HttpField("Test", "three, four"));
         assertThat(fields.stream().map(HttpField::toString).collect(Collectors.toList()), contains("Test: one, two, three, four"));
     }
+
+    @Test
+    public void testWrapperComputeFieldCallingOnField()
+    {
+        var wrapper = new HttpFields.Mutable.Wrapper(HttpFields.build())
+        {
+            final List<String> actions = new ArrayList<>();
+
+            @Override
+            public HttpField onAddField(HttpField field)
+            {
+                actions.add("onAddField");
+                return super.onAddField(field);
+            }
+
+            @Override
+            public boolean onRemoveField(HttpField field)
+            {
+                actions.add("onRemoveField");
+                return super.onRemoveField(field);
+            }
+
+            @Override
+            public HttpField onReplaceField(HttpField oldField, HttpField newField)
+            {
+                actions.add("onReplaceField");
+                return super.onReplaceField(oldField, newField);
+            }
+        };
+
+        wrapper.computeField("non-existent", (name, httpFields) -> null);
+        assertThat(wrapper.size(), is(0));
+        assertThat(wrapper.actions, is(List.of()));
+
+        wrapper.computeField("non-existent", (name, httpFields) -> new HttpField("non-existent", "a"));
+        wrapper.computeField("non-existent", (name, httpFields) -> new HttpField("non-existent", "b"));
+        wrapper.computeField("non-existent", (name, httpFields) -> null);
+        assertThat(wrapper.size(), is(0));
+        assertThat(wrapper.actions, is(List.of("onAddField", "onReplaceField", "onRemoveField")));
+        wrapper.actions.clear();
+
+        wrapper.computeField(HttpHeader.VARY, (name, httpFields) -> null);
+        assertThat(wrapper.size(), is(0));
+        assertThat(wrapper.actions, is(List.of()));
+
+        wrapper.computeField(HttpHeader.VARY, (name, httpFields) -> new HttpField(HttpHeader.VARY, "a"));
+        wrapper.computeField(HttpHeader.VARY, (name, httpFields) -> new HttpField(HttpHeader.VARY, "b"));
+        wrapper.computeField(HttpHeader.VARY, (name, httpFields) -> null);
+        assertThat(wrapper.size(), is(0));
+        assertThat(wrapper.actions, is(List.of("onAddField", "onReplaceField", "onRemoveField")));
+        wrapper.actions.clear();
+    }
 }

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpGeneratorClientTest.java
@@ -99,8 +99,8 @@ public class HttpGeneratorClientTest
 
         HttpFields.Mutable fields = HttpFields.build();
         fields.add("Host", "something");
-        assertThrows(IllegalArgumentException.class, () -> fields.add("Null", (String)null));
-        assertThrows(IllegalArgumentException.class, () -> fields.add("Null", (List<String>)null));
+        assertThrows(NullPointerException.class, () -> fields.add("Null", (String)null));
+        assertThrows(NullPointerException.class, () -> fields.add("Null", (List<String>)null));
         fields.add("Empty", "");
         RequestInfo info = new RequestInfo("GET", "/index.html", fields);
         assertFalse(gen.isChunking());

--- a/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ServletContextHandlerTest.java
+++ b/jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ServletContextHandlerTest.java
@@ -79,6 +79,7 @@ import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.LocalConnector;
 import org.eclipse.jetty.server.Request;
 import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.ResponseUtils;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.handler.ContextHandler;
 import org.eclipse.jetty.server.handler.ContextHandlerCollection;
@@ -103,6 +104,7 @@ import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
@@ -690,6 +692,31 @@ public class ServletContextHandlerTest
     {
         _server.stop();
         _server.join();
+    }
+
+    @Test
+    public void testEnsureNotPersistent() throws Exception
+    {
+        ServletContextHandler root = new ServletContextHandler("/", ServletContextHandler.SESSIONS);
+        root.setContextPath("/");
+        root.addServlet(new ServletHolder(new HttpServlet()
+        {
+            @Override
+            protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            {
+                Request request = ((ServletApiRequest)req).getRequest();
+                Response response = ((ServletApiResponse)resp).getResponse();
+
+                ResponseUtils.ensureNotPersistent(request, response);
+            }
+        }), "/ensureNotPersistent");
+        _server.setHandler(root);
+
+        _server.start();
+
+        String rawResponse = _connector.getResponse("GET /ensureNotPersistent HTTP/1.0\r\n\r\n");
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+        assertThat(response.getStatus(), is(200));
     }
 
     @Test

